supabase/migrations/20240510123247_allow_anonymising_client.sql

```diff
@@ -0,0 +1,119 @@
+    clients.phone_number,
+    clients.is_active,
+    family_count.family_count
+   FROM (clients
+     LEFT JOIN family_count ON ((clients.family_id = family_count.family_id)))
+  ORDER BY clients.full_name;
+
+
+create or replace view "public"."parcels_plus" as  SELECT parcels.primary_key AS parcel_id,
+    parcels.collection_datetime,
+    parcels.packing_date,
+    parcels.created_at,
+    packing_slots.name AS packing_slot_name,
+    packing_slots."order" AS packing_slot_order,
+    parcels.voucher_number,
+    collection_centres.name AS collection_centre_name,
+    collection_centres.acronym AS collection_centre_acronym,
+    clients.primary_key AS client_id,
+    clients.full_name AS client_full_name,
+    clients.address_postcode AS client_address_postcode,
+    clients.flagged_for_attention AS client_flagged_for_attention,
+    clients.signposting_call_required AS client_signposting_call_required,
+    clients.phone_number AS client_phone_number,
+    clients.is_active AS client_is_active,
+    family_count.family_count,
+    last_status.event_name AS last_status_event_name,
+    last_status.event_data AS last_status_event_data,
+    last_status."timestamp" AS last_status_timestamp,
+    last_status.workflow_order AS last_status_workflow_order
+   FROM (((((parcels
+     LEFT JOIN collection_centres ON ((parcels.collection_centre = collection_centres.primary_key)))
+     LEFT JOIN clients ON ((parcels.client_id = clients.primary_key)))
+     LEFT JOIN packing_slots ON ((parcels.packing_slot = packing_slots.primary_key)))
+     LEFT JOIN family_count ON ((family_count.family_id = clients.family_id)))
+     LEFT JOIN last_status ON ((last_status.parcel_id = parcels.primary_key)))
+  ORDER BY parcels.packing_date DESC;
+
+set check_function_bodies = off;
+
+CREATE OR REPLACE FUNCTION public."deactivateClient"(clientid uuid)
+ RETURNS void
+ LANGUAGE plpgsql
+AS $function$DECLARE
+    updated_family_id UUID;
+BEGIN  
+    UPDATE clients 
+    SET 
+        full_name = null,
+        phone_number = null,
+        address_1 = null,
+        address_2 = null,
+        address_town = null,
+        address_county = null,
+        address_postcode = null,
+        delivery_instructions = null,
+        dietary_requirements = null,
+        feminine_products = null,
+        baby_food = null,
+        pet_food = null,
+        other_items = null,
+        extra_information = null,
+        flagged_for_attention = null,
+        signposting_call_required = null,
+        is_active = false
+    WHERE 
+        primary_key = clientId;
+
+    SELECT family_id INTO updated_family_id
+    FROM clients
+    WHERE primary_key = clientId;
+
+    DELETE FROM families WHERE family_id = updated_family_id;
+
+END;$function$
+;
+
+create policy "Disable updates on inactive clients"
+on "public"."clients"
+as restrictive
+for update
+to authenticated
+using ((is_active = true))
+with check (((is_active = true) OR (is_active = false)));
```

Comment:
These migration files are automatically created right? Just curious why does it say `(is_active = true) OR (is_active = false)` here

Reply:
the migration files are typically automatically created, although it is possible to write the SQL manually. in this case it was automatically generated. however, I wrote the with check expression myself. 
The purpose of this RLS policy is to disable editing an inactive client, hence the `using ((is_active = true))`. the using expression applies to the old value of the field, before the update.
The with check expression applies to the new value of the field, after the update. 
If there's no with check expression, the "using" expression is also used for the "with check", but applied to the new value. 
In our case, we want to disable editing an inactive client, but we want to allow the value of `is_active` to be either true or false afterwards. We need to include the with check to avoid "is_active" also being checked to be true after the update, which would prevent deleting clients altogether.
See https://supabase.com/docs/guides/database/postgres/row-level-security for more info

Reply:
I see, thank you for explaining :) So it runs this policy on the clients table every time there is an update, the request comes from an authenticated entity and we can only change rows where `is_active == true`, and we can change that to either true or false, is that right? Not entirely sure what restrictive does

Also is there a reason for the many brackets? Does it error if you only have `using (is_active = true)` and `with check ((is_active = true) OR (is_active = false))`?

Reply:
Yes that's right! Restrictive means its combined with the other policy on this table (which is only allowing authenticated users to make changes) in such a way that both policies have to be satisfied in order for the change to be allowed. The other option is permissive, in which case only the policy itself has to be satisified in order for the change to be allowed.

supabase added the additional brackets once I added the code into the RLS policy. I just tried to remove them and reset the database, and it seemed to work fine. So I think we should be ok to remove the additional brackets.